src/SqlBinding/TriggerBinding/SqlTableWatcher.cs

```diff
@@ -0,0 +1,505 @@
+        /// <exception cref="ArgumentNullException">
+        /// Thrown if any of the parameters are null
+        /// </exception>
+        public SqlTableWatcher(string table, string connectionStringSetting, IConfiguration configuration, ITriggeredFunctionExecutor executor)
+        {
+            _table = table ?? throw new ArgumentNullException(nameof(table));
+            _connectionStringSetting = connectionStringSetting ?? throw new ArgumentNullException(nameof(connectionStringSetting));
+            _configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+            _executor = executor ?? throw new ArgumentNullException(nameof(executor));
+            _workerTable = "Worker_Table_" + _table;
+
+            _cancellationTokenSource = new CancellationTokenSource();
+            _rows = new List<Dictionary<string, string>>();
+            _queryStrings = new Dictionary<string, string>();
+            _primaryKeys = new Dictionary<string, string>();
+            _whereChecksOfRows = new Dictionary<Dictionary<string, string>, string>();
+            _primaryKeyValuesOfRows = new Dictionary<Dictionary<string, string>, string>();
+
+            _checkForChangesTimer = new Timer(CheckForChangesCallback);
+            _renewLeasesTimer = new Timer(RenewLeasesCallback);
+            // Should these be just normal semaphores?
+            _checkForChangesLock = new SemaphoreSlim(1);
+            _renewLeasesLock = new SemaphoreSlim(1);
+
+        }
+
+        /// <summary>
+        /// Starts the watcher which begins polling for changes on the user's table specified in the constructor
+        /// </summary>
+        /// <returns></returns>
+        public async Task StartAsync()
+        {
+            await CreateWorkerTableAsync();
+            _state = State.CheckingForChanges;
+            _checkForChangesTimer.Change(0, _pollingInterval * 1000);
+            _renewLeasesTimer.Change(0, _leaseTime * 1000);
+        }
+
+        /// <summary>
+        /// Stops the watcher which stops polling for changes on the user's table
+        /// </summary>
+        /// <returns></returns>
+        public async Task StopAsync()
+        {
+            await _checkForChangesTimer.DisposeAsync();
+            await _renewLeasesTimer.DisposeAsync();
+        }
+
+        /// <summary>
+        /// Executed once every "_leastTime" period. If the state of the watcher is <see cref="State.ProcessingChanges"/>, then 
+        /// we will renew the leases held by the watcher on "_rows"
+        /// </summary>
+        /// <param name="state">Unused </param>
+        private async void RenewLeasesCallback(object state)
+        {
+            await _renewLeasesLock.WaitAsync();
+            try
+            {
+                // Should I be using the state parameter instead? Is it unsafe otherwise because the other timer thread can modify this instance variable?
+                if (_state == State.ProcessingChanges)
```

Comment:
Yeah, I guess this is what I was getting at in this comment [here](https://github.com/Azure/azure-functions-sql-extension/blob/triggerbinding/src/SqlBinding/TriggerBinding/SqlTableWatcher.cs#L165). I definitely need to add something that recognizes if the change processing thread keeps trying to execute the function and failing so that it eventually knows to give up and let another worker try. I'm not totally how a thread could get "stuck" in trying to execute the function/some SQL command, though. Is this possible? In that case, just having a count for how many failed function executions the thread has experienced wouldn't be enough.